EventKit/eventkit-cloud
# Clean up stuck celery tasks

## Description of Improvement
Celery has a built in mechanism to shutdown tasks that run too long. But in some cases this isn't working, so we need a way to kill celery tasks that are unresponsive to the normal shutdown method. 

This adds a scheduled task that runs periodically and checks for tasks that have been running too long and need to be killed. It kills those tasks, first trying to gracefully stop them, then hard killing them. It also modifies the state of the run, dataprovidertaskrecord, and exporttaskrecord such that the stuck tasks are canceled, any pending/waiting tasks are left alone, the dataprovidertaskrecord is pending, and the run is submitted. This way the run is able to be picked back up by a new worker than can complete any remaining, non-stuck, work. 

## How to test
First, you'll need to make a couple changes. The configuration in this PR is intended for production, not test. 
You'll need to:

Add `TASK_TIMEOUT` to your `.env`. It should be an integer number of seconds that a task is allowed to run before killing it. 
I'd recommend something around 60 seconds, such that testing is fast. 

Comment out this line: https://github.com/EventKit/eventkit-cloud/compare/worker_hanging_on_tiffs#diff-7d676b6ab86d097fa11139f76d569eaaf19007273358d08b818bef07c130f401R104 
Otherwise Celery's default shutdown mechanism will shut down tasks before the mechanism in this PR acts. 

Add `TASK_TIMEOUT` to the environment for eventkit and celery in the docker-compose.yml. 

Make `clean-up-stuck-tasks` run more frequently, once again something in the ballpark of 60 seconds seems wise. This can be done by changing the time here: https://github.com/EventKit/eventkit-cloud/compare/worker_hanging_on_tiffs#diff-7d676b6ab86d097fa11139f76d569eaaf19007273358d08b818bef07c130f401R48
or in the admin console under `Periodic Tasks`

After completing all of this - make sure it is reflected in your running instance of eventkit. I.e. `make down && make up` if needed. 

Then create and run an export with (at least) one lightweight dataset and one heavyweight dataset. Ideally what we're shooting for is an export in which one of the providers will have finished before `TASK_TIMEOUT` is up, and the other will *not* have finished. 

Watch the export run, verify via the logs that when `clean-up-stuck-tasks` runs after the export has been running for `task_timeout` seconds or longer that the tasks that are still running are marked as cancelled. The run should be marked as submitted for a bit. The celery worker running the long running tasks should be killed - this may take a moment as we wait for the kill to escalate from soft kill to hard kill. THEN the run should get picked up by a new worker. The final state of the run should be 'incomplete'. The 'stuck' tasks should be cancelled. Any tasks that had already finished should show that they're finished. And there should be a ZIP available for download, containing only the data from the completed (non-stuck/non-cancelled) tasks. A

## Quality Assurance 

The following items have been updated:
 
## Screenshots



:smile:
